backend/FwLite/MiniLcm/SyncHelpers/DiffCollection.cs

```diff
@@ -1,141 +1,267 @@
-            await replace(api, createdItem, afterItem);
+            await diffApi.Replace(createdItem, afterItem);
         }
 
         return changes;
     }
 
-    /// <summary>
-    ///
-    /// </summary>
-    /// <param name="api"></param>
-    /// <param name="before"></param>
-    /// <param name="after"></param>
-    /// <param name="identity"></param>
-    /// <param name="add"></param>
-    /// <param name="remove"></param>
-    /// <param name="replace">api, before, after is the parameter order</param>
-    /// <typeparam name="T"></typeparam>
-    /// <typeparam name="TId"></typeparam>
-    /// <returns></returns>
     public static async Task<int> Diff<T, TId>(
-        IMiniLcmApi api,
         IList<T> before,
         IList<T> after,
-        Func<T, TId> identity,
-        Func<IMiniLcmApi, T, Task<int>> add,
-        Func<IMiniLcmApi, T, Task<int>> remove,
-        Func<IMiniLcmApi, T, T, Task<int>> replace) where TId : notnull
+        CollectionDiffApi<T, TId> diffApi) where TId : notnull
     {
         var changes = 0;
-        var afterEntriesDict = after.ToDictionary(identity);
+        var afterEntriesDict = after.ToDictionary(diffApi.GetId);
         foreach (var beforeEntry in before)
         {
-            if (afterEntriesDict.TryGetValue(identity(beforeEntry), out var afterEntry))
+            if (afterEntriesDict.TryGetValue(diffApi.GetId(beforeEntry), out var afterEntry))
             {
-                changes += await replace(api, beforeEntry, afterEntry);
+                changes += await diffApi.Replace(beforeEntry, afterEntry);
             }
             else
             {
-                changes += await remove(api, beforeEntry);
+                changes += await diffApi.Remove(beforeEntry);
             }
 
-            afterEntriesDict.Remove(identity(beforeEntry));
+            afterEntriesDict.Remove(diffApi.GetId(beforeEntry));
         }
 
         foreach (var value in afterEntriesDict.Values)
         {
-            changes += await add(api, value);
+            changes += await diffApi.Add(value);
         }
 
         return changes;
     }
 
-    public static async Task<int> Diff<T>(
-        IMiniLcmApi api,
+    public static async Task<int> DiffOrderable<T>(
         IList<T> before,
         IList<T> after,
-        Func<IMiniLcmApi, T, Task<int>> add,
-        Func<IMiniLcmApi, T, Task<int>> remove,
-        Func<IMiniLcmApi, T, T, Task<int>> replace) where T : IObjectWithId
+        OrderableCollectionDiffApi<T> diffApi) where T : IOrderable
     {
-        return await Diff(api, before, after, t => t.Id, add, remove, replace);
+        var changes = 0;
+
+        var positionDiffs = DiffPositions(before, after);
+        if (positionDiffs is not null)
+        {
+            var stableIds = after.Where((_, i) => !positionDiffs.ContainsKey(i))
```

Comment:
key here is the SortIndex on PositionDiff, meaning removes are negative which will never match here, is that ok? I'm not sure but it feels like a bug, I guess stableIds aren't used for removes so maybe it's ok?

Reply:
I'll add comments to clarify this. But, yes it's intentional that this line doesn't pick up removed items.
"Stable IDs" are all the IDs in `after` that weren't added or moved. The **positive** keys in positionDiffs are precisely the after/current indexes that were added/moved i.e. the unstable indexes.